### PR TITLE
query object pattern over queries for object in model or controller

### PR DESCRIPTION
**Pull request by kamalpandey**
https://github.com/BiggerPockets/biggerpockets/pull/8603

### CODE REVIEW

File: rails/query-object-over-queries-on-models-or-controllers.md [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+
+    def initialize(base_relation=nil)
+      base_relation = WordpressBlogPost.includes(:wordpress_blog_post_category, :wordpress_author) unless base_relation
+    end
+
+    def call
+      relation.where.not(wordpress_blog_post_category_id: WordpressBlogPostCategory.
+        invalid_wordpress_blog_post_category.select(:id)).
+        order(created_at: :desc).
+        references(:wordpress_blog_post_category)
```
Comment (lewispb) [inline rails/query-object-over-queries-on-models-or-controllers.md]: Seeing it here, I'm not sure about the call pattern to be honest 🤔 What does everyone think?

File: rails/query-object-over-queries-on-models-or-controllers.md [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+
+A better Alternative is to use the Query Object.
+
+
+````ruby
+module WordpressBlogPosts
+  class RecentPostsQuery
+    class << self
+      delegate :call, to: :new
+    end
```
Comment (lewispb) [inline rails/query-object-over-queries-on-models-or-controllers.md]: I guess if we did go this route we could abstract this to a base class, `ApplicationQuery` which might not be a bad idea anyway.

File: rails/query-object-over-queries-on-models-or-controllers.md [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+    includes(:wordpress_blog_post_category, :wordpress_author).
+      where.not(wordpress_blog_post_category_id: WordpressBlogPostCategory.
+      invalid_wordpress_blog_post_category.select(:id)).
+      order(created_at: :desc).
+      references(:wordpress_blog_post_category).
+  end
+end
+````
+
+A better Alternative is to use the Query Object.
```
Comment (lewispb) [inline rails/query-object-over-queries-on-models-or-controllers.md]:
```suggestion
A better alternative is to use the Query Object.
```

File: rails/query-object-over-queries-on-models-or-controllers.md (hunk 1 above)
Comment (bhanubhakta) [reply rails/query-object-over-queries-on-models-or-controllers.md]: I agree with @lewispb . I think the query should only be dependent on the passed relation. In contrast, this query includes `WordpressBlogPostCategory.invalid_wordpress_blog_post_category.select(:id)`.

Comment (bhanubhakta) [reply rails/query-object-over-queries-on-models-or-controllers.md]: I am open to discussions here.

File: rails/query-object-over-queries-on-models-or-controllers.md [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+  end
+end
+
+module WordpressBlogPosts
+  class RecentPostsQuery < ApplicationQuery
+
+    attr_reader :relation
+
+    def initialize(base_relation=nil)
+      base_relation = WordpressBlogPost.includes(:wordpress_blog_post_category, :wordpress_author) unless base_relation
```
Comment (lewispb) [inline rails/query-object-over-queries-on-models-or-controllers.md]:
```suggestion
      @relation ||= WordpressBlogPost.includes(:wordpress_blog_post_category, :wordpress_author) unless base_relation
```

File: rails/query-object-over-queries-on-models-or-controllers.md (hunk 1 above)
Comment (kxhitiz) [reply rails/query-object-over-queries-on-models-or-controllers.md]: using call pattern is too much rails magic?

Comment (kamalpandey) [reply rails/query-object-over-queries-on-models-or-controllers.md]: @lewispb @kxhitiz I have updated the the code to extract the call pattern to the ApplicationQuery based on the feedback below. What do you think about it?

File: rails/query-object-over-queries-on-models-or-controllers.md [hunk 5]
```diff
@@ -46,34 +46,28 @@ end
...
 ````ruby
-class RecentWordpressBlogPostQuery < ApplicationQuery
-
-  def initialize(relation = nil)
-    @relation = relation 
+class ApplicationQuery
+  def self.all
+    new.all
   end
+end
```
Comment (lewispb) [inline rails/query-object-over-queries-on-models-or-controllers.md]: Thanks @kamalpandey - maybe we don't need to include this ApplicationQuery?

Comment (kamalpandey) [reply rails/query-object-over-queries-on-models-or-controllers.md]: :+1:

**Pull request merged**